helm: document ipv{4,6}NativeRoutingCIDR option missing in helm reference

vincentmli opened the pull request:
helm options that are commented out in install/kubernetes/cilium/values.yaml.tmpl
are missing documentation in Documentation/helm-values.rst and in helm reference
this is reported in a few GH issues, for example:
https://github.com/cilium/cilium/issues/21107

This appears to be known helm-doc bug and the bug has not been resolved, a related
PR https://github.com/norwoodj/helm-docs/pull/99 that seems has not completely resloved
the issue.

before the helm-doc bug is root caused, uncomment these helm options as workaround.

This PR is to document ipv4, ipv6 native routing cidr helm option

Signed-off-by: Vincent Li <v.li@f5.com>

Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Thanks for contributing!

<!-- Description of change -->

Fixes: #21107

```release-note
document ipv4/ipv6 native routing cidr helm option missing in Documentation and helm reference
```

sayboras commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -307,31 +307,31 @@ bpf:
   # host stack (true) or directly and more efficiently out of BPF (false) if
   # the kernel supports it.
-  #hostRouting: true
+  hostRouting: true
```
not sure if we want to backport this change, if not, we can just remove this deprecated field outright

sayboras commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -656,13 +656,13 @@ hubble:
   # -- Buffer size of the channel Hubble uses to receive monitor events. If this
   # value is not set, the queue size is set to the default monitor queue size.
-  # eventQueueSize: ""
+  eventQueueSize: ""
```
just to avoid parsing error

```suggestion
  # -- Buffer size of the channel Hubble uses to receive monitor events. If this
  # value is not set or zero, the queue size is set to the default monitor queue size.
  eventQueueSize: "0"
```

vincentmli replied:
got it

sayboras commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -387,12 +387,12 @@ cni:
   # configuration outside of a Kubernetes environment. This parameter is
   # mutually exclusive with the 'cni.configMap' parameter.
-  # readCniConf: /host/etc/cni/net.d/05-cilium.conf
+  readCniConf: /host/etc/cni/net.d/05-cilium.conf
```
to avoid file not found upon agent start

```suggestion
  # Example value: /host/etc/cni/net.d/05-cilium.conf
  readCniConf: ""
```

sayboras reviewed (commented):
Thanks a lot for your PR, I have couple of comments below. 

One of them will help with the below smoke test failure you mentioned. I would also suggest taking a quick look in code to confirm the default value in helm and implementation are the same, sometimes we didn't pay much attention in helm especially it's commented.

qmonnet commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -420,7 +420,7 @@ containerRuntime:
   integration: none
   # -- Configure the path to the container runtime control socket.
-  # socketPath: /path/to/runtime.sock
+  socketPath: /path/to/runtime.sock
```
This one probably needs some adjustment as well.

vincentmli replied:
probably before I can actually test it, maybe I should not change it :)

squeed commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -387,12 +387,12 @@ cni:
   # interpret the cniConf variable as CNI configuration file and write it
   # when the agent starts up
-  # configMap: cni-configuration
+  configMap: cni-configuration
```
This is not safe to set by default.

vincentmli replied:
I will not uncomment it for now since I am not sure what this helm option means and how to set it.

gandro commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -20,9 +20,13 @@ rbac:
-# k8sServicePort:
+# -- (string) Kubernetes config path
+# @default -- ~/.kube/config
+kubeConfigPath: ~
```
Since we're using the `if value` pattern here, I think using empty string (`""`) instead of `~` for those string values is more straight forward.

Let's only use nil for values where we do an explicit nil check.

vincentmli replied:
I thought `if value` is to do `nil` check, but I guess you mean ` != nil` or `== nil` is to do `nil` check in `pkg/option/config.go`? can you give an example ? :)

gandro replied:
Yes, ``{{ if value }}`` does a nil check. But what I'm proposing is that you set all string types to empty string instead of nil, because `""` will behave exactly like "~" for `{{ if value }}`. 

For example:

```suggestion
kubeConfigPath: ""
```

The `nil` values are mostly useful where `{{ if value }}` cannot be used (e.g. everything in your third commit).

vincentmli replied:
I think I got it, thanks!

gandro commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -267,23 +267,27 @@ bpf:
+  # -- (int) Configure the maximum number of entries in the TCP connection tracking
   # table.
-  # ctTcpMax: '524288'
+  # @default -- '524288'
```
Can you change the quotes of all default values to use ``` instead? It seems that way the formatting is consistent with what helm-docs does without `@default`. Thanks.

vincentmli replied:
seems your comment missed something :) I see use `instead?`, maybe you mean remove the single quote `'`?

gandro replied:
Yeah, sorry, the formatting was of. What I meant is that you replace the single quote `'` with a tick marks. For example:
```suggestion
  # @default -- `524288`
```

gandro commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -247,7 +247,7 @@ data:
 {{- end }}
 
-{{- if hasKey .Values.bpf "tproxy" }}
+{{- if and (hasKey .Values.bpf "tproxy") (not (kindIs "invalid" .Values.bpf.tproxy)) }}
```
I know I proposed this pattern, but now I wonder: Would the following actually be sufficient?

```suggestion
{{- if (not (kindIs "invalid" .Values.bpf.tproxy)) }}
```

vincentmli replied:
ok I will try that, thanks!

gandro commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1159,8 +1154,9 @@ ipMasqAgent:
-# iptablesLockTimeout defines the iptables "--wait" option when invoked from Cilium.
-# iptablesLockTimeout: "5s"
+# -- (string) iptablesLockTimeout defines the iptables "--wait" option when invoked from Cilium.
+# @default -- `5s`
```
```suggestion
# @default -- `"5s"`
```

vincentmli replied:
ack

gandro commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1203,10 +1199,11 @@ readinessProbe:
 # Valid options are "disabled", "partial", "strict".
 # ref: https://docs.cilium.io/en/stable/gettingstarted/kubeproxy-free/
-#kubeProxyReplacement: "disabled"
+# @default -- `disabled`
```
```suggestion
# @default -- `"disabled"`
```

gandro commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1464,7 +1478,7 @@ tls:
 
 # -- Configure VXLAN and Geneve tunnel port.
-# @default -- Port 8472 for VXLAN, Port 6081 for Geneve
+# @default -- `Port 8472 for VXLAN, Port 6081 for Geneve`
```
This is not an actual value, let's not use the ticks here (i.e. please undo this change).

gandro replied:
This has net yet been resolved.

vincentmli replied:
I am trying to resolve conflicts so you may not see the change, hm, maybe I should stop messing with conflicts now and just address the review, then later address the conflicts since I may need some guidance on that anyway

gandro reviewed (commented):
Thanks for the effort, I know maintaining this is a pain. A few more remaining comments regarding `@default`

gandro commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -309,32 +312,31 @@ bpf:
+  # -- (list) Configure explicitly allowed VLAN id's for bpf logic bypass.
   # [0] will allow all VLAN id's without any filtering.
-  # vlanBypass: []
+  # @default -- `[0]`
```
The default here actually is `[]`

vincentmli replied:
ack

gandro reviewed (commented):
Looks good. Two remaining issues regarding the default values that I spotted.

vincentmli committed:
helm: document ipv{4,6}NativeRoutingCIDR option missing in helm reference

helm options that are commented out in install/kubernetes/cilium/values.yaml.tmpl
are missing documentation in Documentation/helm-values.rst and in helm reference
this is reported in a few GH issues, for example:
https://github.com/cilium/cilium/issues/21107

This appears to be known helm-doc bug and the bug has not been resolved, a related
PR https://github.com/norwoodj/helm-docs/pull/99 that seems has not completely resloved
the issue.

before the helm-doc bug is root caused, uncomment these helm options as workaround.

Since Documentation/helm-values.rst install/kubernetes/cilium/README.md are auto
generated, it is difficult to avoid rebase/squash conflicts in local branch and
upstream master branch, so introduce the helm option in smaller changes so it easy
to review and rebase/squash and merge.

Fixes: #21334 #21107

Signed-off-by: Vincent Li <v.li@f5.com>

sayboras reviewed (approved):
Thanks a lot 💯

gandro reviewed (approved):
Looks good then, thanks! Could you update the PR title to refer to the native routing cidr for the release notes?
